internal/pubsub/subscriber.go

```diff
@@ -0,0 +1,372 @@
+		c.logger = l
+	}
+}
+
+// WithResourceTypeNames sets the resource type names for the client to listen for.
+func WithResourceTypeNames(typeNames []string) ClientOpt {
+	return func(c *Client) {
+		c.resourceTypeNames = typeNames
+	}
+}
+
+// WithQueryEngine sets the query engine for the client.
+func WithQueryEngine(e query.Engine) ClientOpt {
+	return func(c *Client) {
+		c.qe = e
+	}
+}
+
+// WithConn reuses a NATS connection for the client instead of creating one in NewClient.
+func WithConn(nc *nats.Conn) ClientOpt {
+	return func(c *Client) {
+		c.nc = nc
+	}
+}
+
+func defaultLogger() *zap.SugaredLogger {
+	return zap.NewNop().Sugar()
+}
+
+func defaultConn(cfg Config) (*nats.Conn, error) {
+	natsOpts := []nats.Option{
+		nats.Name(cfg.Name),
+		nats.UserCredentials(cfg.Credentials),
+		nats.DrainTimeout(drainTimeout),
+	}
+
+	nc, err := nats.Connect(cfg.Server, natsOpts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return nc, nil
+}
+
+// NewClient creates a new pubsub client.
+func NewClient(cfg Config, opts ...ClientOpt) (*Client, error) {
+	var c Client
+
+	for _, opt := range opts {
+		opt(&c)
+	}
+
+	// If we don't have an existing NATS connection, create one
+	if c.nc == nil {
+		nc, err := defaultConn(cfg)
+		if err != nil {
+			return nil, err
+		}
+
+		c.nc = nc
+	}
+
+	// If we don't have a logger, use a nop logger.
+	if c.logger == nil {
+		c.logger = defaultLogger()
+	}
+
+	js, err := c.nc.JetStream()
+	if err != nil {
+		return nil, err
+	}
+
+	c.js = js
+	c.stream = cfg.Stream
+	c.consumer = cfg.Consumer
+	c.prefix = cfg.Prefix
+
+	return &c, nil
+}
+
+func (c *Client) ensureStream() error {
+	c.logger.Debugw("checking that NATS stream exists", "stream_name", c.stream)
+
+	_, err := c.js.StreamInfo(c.stream)
+	if err == nil {
+		c.logger.Debugw("stream exists, not recreating", "stream_name", c.stream)
+		return nil
+	}
+
+	if !errors.Is(err, nats.ErrStreamNotFound) {
+		return err
+	}
+
+	_, err = c.js.AddStream(&nats.StreamConfig{
+		Name: c.stream,
+		Subjects: []string{
+			c.prefix + ".>",
+		},
+		Storage:   nats.FileStorage,
+		Retention: nats.LimitsPolicy,
+		Discard:   nats.DiscardNew,
+	})
+
+	return err
+}
+
+// Listen ensures a stream exists, binds to it, and listens for resource lifecycle events on that stream.
+func (c *Client) Listen() error {
+	// Ensure stream exists before we attempt to listen on it
+	err := c.ensureStream()
+	if err != nil {
+		return err
+	}
+
+	// Set subscription options. We specifically want manual acks in case something goes wrong
+	// persisting the event
+	subOpts := []nats.SubOpt{
+		nats.BindStream(c.stream),
+		nats.Durable(c.consumer),
+		nats.ManualAck(),
+		nats.AckExplicit(),
+	}
+
+	for _, name := range c.resourceTypeNames {
+		subject := fmt.Sprintf("%s.%s.>", c.prefix, name)
+		queueName := fmt.Sprintf("%s-%s", c.consumer, name)
+
+		subscription, err := c.js.QueueSubscribe(subject, queueName, c.receiveMsg, subOpts...)
+		if err != nil {
+			return err
+		}
```

Comment:
This will process each subscription serially btw - which might be what we want, just pointing it out.

Reply:
Good call-out. I think that's fine for now, since we don't have a good sense for what our resource limits will be or whether certain subjects will require more resources to process than others. Leaving this at a single subscription per resource type is probably better than trying to optimize for things we don't know yet.